x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/bucket/geogrid/AbstractGeoHexGridTiler.java

```diff
@@ -0,0 +1,190 @@
+            if (minH3 == maxH3) {
+                return setValuesFromPointResolution(minH3, values, geoValue);
+            }
+            // TODO: specialize when they are neighbour cells.
+        }
+        // recurse tree
+        return setValuesByRecursion(values, geoValue, bounds);
+    }
+
+    /**
+     * It calls {@link #maybeAdd(long, GeoRelation, GeoShapeCellValues, int)} for {@code h3} and the neighbour cells if necessary.
+     */
+    private int setValuesFromPointResolution(long h3, GeoShapeCellValues values, GeoShapeValues.GeoShapeValue geoValue) throws IOException {
+        int valueIndex = 0;
+        {
+            final GeoRelation relation = relateTile(geoValue, h3);
+            valueIndex = maybeAdd(h3, relation, values, valueIndex);
+            if (relation == GeoRelation.QUERY_CONTAINS) {
+                return valueIndex;
+            }
+        }
+        // Point resolution is done using H3 library which uses spherical geometry. It might happen that in cartesian, the
+        // actual point value is in a neighbour cell as well.
+        {
+            for (long n : H3.hexRing(h3)) {
+                final GeoRelation relation = relateTile(geoValue, n);
+                valueIndex = maybeAdd(n, relation, values, valueIndex);
+                if (relation == GeoRelation.QUERY_CONTAINS) {
+                    return valueIndex;
+                }
+            }
+        }
+        return valueIndex;
+    }
+
+    /**
+     * Adds {@code h3} to {@link GeoShapeCellValues} if {@link #relateTile(GeoShapeValues.GeoShapeValue, long)} returns
+     * a relation different to {@link GeoRelation#QUERY_DISJOINT}.
+     */
+    private int maybeAdd(long h3, GeoRelation relation, GeoShapeCellValues values, int valueIndex) throws IOException {
+        if (relation != GeoRelation.QUERY_DISJOINT) {
+            values.resizeCell(valueIndex + 1);
+            values.add(valueIndex++, h3);
+        }
+        return valueIndex;
+    }
+
+    /**
+     * Recursively search the H3 tree, only following branches that intersect the geometry.
+     * Once at the required depth, then all cells that intersect are added to the collection.
+     */
+    // package private for testing
+    int setValuesByRecursion(GeoShapeCellValues values, GeoShapeValues.GeoShapeValue geoValue, GeoShapeValues.BoundingBox bounds)
+        throws IOException {
+        // NOTE: When we recurse, we cannot shortcut for CONTAINS relationship because it might fail when visiting noChilds.
+        int valueIndex = 0;
+        if (bounds.maxX() - bounds.minX() < 180d) {
+            final long minH3 = H3.geoToH3(bounds.minY(), bounds.minX(), 0);
+            final long maxH3 = H3.geoToH3(bounds.maxY(), bounds.maxX(), 0);
+            if (minH3 == maxH3) {
+                valueIndex = setValuesByRecursion(values, geoValue, minH3, 0, valueIndex);
+                for (long n : H3.hexRing(minH3)) {
```

Comment:
Maybe we can skip it if the relationship is CONTAINS but we should test it, it would be a nice follow up. I don't care so much here because we are caching the points at that level so it should not be too expensive.